Fix a job quota related deadlock

ASverdlov opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
In case ResourceQuota is used and sets a max # of jobs, a CronJob may get trapped in a deadlock:
  1. Job quota for a namespace is reached.
  2. CronJob controller can't create a new job, because quota is reached.
  3. Cleanup of jobs owned by a cronjob doesn't happen, because a control loop iteration is finished because of an error to create a job.

To fix this we stop early quitting from a control loop iteration when cronjob reconciliation failed and always let old jobs to be cleaned up.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes [#119775](https://github.com/kubernetes/kubernetes/issues/119775)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed an issue where a CronJob could fail to clean up Jobs when the ResourceQuota for Jobs had been reached.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ASverdlov committed:
Fix a job quota related deadlock

In case ResourceQuota is used and sets a max # of jobs, a CronJob may get
trapped in a deadlock:
  1. Job quota for a namespace is reached.
  2. CronJob controller can't create a new job, because quota is
     reached.
  3. Cleanup of jobs owned by a cronjob doesn't happen, because a
     control loop iteration is finished because of an error to create a
     job.

To fix this we stop early quitting from a control loop iteration when
cronjob reconciliation failed and always let old jobs to be cleaned up.

aojea commented on pkg/controller/cronjob/cronjob_controllerv2.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"time"
 
 	"github.com/robfig/cron/v3"
+	"k8s.io/klog/v2"
```
this change is unnecesary

ASverdlov replied:
Fixed

aojea commented on pkg/controller/cronjob/cronjob_controllerv2.go:
```diff
@@ -199,13 +200,6 @@ func (jm *ControllerV2) sync(ctx context.Context, cronJobKey string) (*time.Dura
 	cronJobCopy, requeueAfter, updateStatus, err := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
 	if err != nil {
 		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)
-		if updateStatus {
```
without much experience on this controller this change looks like an important behavior change in the controller

ASverdlov replied:
There is exactly the same if-statement below, which is going to be executed unless there is a panic in the code, as there are no return-statements in-between: https://github.com/kubernetes/kubernetes/blob/16173c23baa3a6a7ad19233c963d165dc7f602d9/pkg/controller/cronjob/cronjob_controllerv2.go#L208-L214

sftim replied:
Could we nevertheless make attempt to update the status here, but with different error handling, and a comment in both places about the situation?

Logging that the status update failed is helpful; it's just the `return` statements that create a problem.

Any other logic refactor that ensures we attempt `cleanupFinishedJobs()` is also fine by me. Ideally we don't drop any observability along with that change.

sftim replied:
Maybe we should set `updateStatus` to true, but defer the update?

I'd like to see a clear comment that shows why this change is right. It might be, but I want the code and PR to explain why.

ASverdlov replied:
We can still remove the copy-paste like this: https://github.com/ASverdlov/kubernetes/commit/675e411ae6958bc6b07e7ae104480e316e403fd4.

Attempt on a clear comment on why it would work follows: let's compare 2 versions: because the following 2 versions of code are equivalent:
```go
	cronJobCopy, requeueAfter, updateStatus, err := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
	if err != nil {
		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)

                COPY_PASTE()
		return nil, err
	}

        COPY_PASTE()

	if requeueAfter != nil {
		logger.V(4).Info("Re-queuing cronjob", "cronjob", klog.KObj(cronJob), "requeueAfter", requeueAfter)
		return requeueAfter, nil
	}
	// this marks the key done, currently only happens when the cronjob is suspended or spec has invalid schedule format
	return nil, nil

```
and

```go
	cronJobCopy, requeueAfter, updateStatus, reconciliationErr := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
	if err != nil {
		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)
	}
        
        COPY_PASTE()

	if requeueAfter != nil {
		logger.V(4).Info("Re-queuing cronjob", "cronjob", klog.KObj(cronJob), "requeueAfter", requeueAfter)
		return requeueAfter, reconciliationErr
	}
	// this marks the key done, currently only happens when the cronjob is suspended or spec has invalid schedule format
	return nil, reconciliationErr

```

* In case `jm.syncCronJob` doesn't return an error, they are line-by-line the same (if you ignore renaming of the error variable.
* In case `jm.syncCronJob` does return an error, the only difference is: latter code also returns a `requeueAfter` in case it was set. But `jm.syncCronJob()` never returns a `requeueAfter != nil` when it returns an error, so in practice it doesn't change the current behaviour.

ASverdlov committed:
Dont reorder imports

aojea commented:
/ok-to-test

code owners should take a look

Thanks

pacoxu commented:
The issue seems to be a valid case.
/lgtm
/cc @soltysh @alaypatel07
for cronjob controller owners

sftim commented:
Changelog suggestion
```diff
-Fix a cronjob getting stuck when jobs quota is reached
+Fixed an issue where a CronJob could fail to clean up Jobs when the ResourceQuota for Jobs had been reached.
```

ASverdlov committed:
Don't stop requeuing on reconciliation error

Previous code only logged the reconciliation error inside jm.sync() and
didn't return the reconciliation error to it's invoker
processNextWorkItem().

Adding a copy-paste back to avoid this issue.

alaypatel07 commented:
apologies for the delay, I'll take a look at this.

soltysh commented:
/priority important-longterm
/triage accepted
/label tide/merge-method-squash

soltysh commented:
/assign

ASverdlov committed:
Remove copy-pasted cleanupFinishedJobs()

Now we always call jm.cleanupFinishedJobs() first and then
jm.syncCronJob().

We also extract cronJobCopy and updateStatus outside jm.syncCronJob
function and pass pointers to them in both jm.syncCronJob and
jm.cleanupFinishedJobs to make delayed updates handling more explicit
and not dependent on the order in which cleanupFinishedJobs and
syncCronJob are invoked.

ASverdlov commented on pkg/controller/cronjob/cronjob_controllerv2.go:
```diff
@@ -435,7 +441,7 @@ func (jm *ControllerV2) syncCronJob(
 			cjCopy, err := jm.cronJobControl.GetCronJob(ctx, cronJob.Namespace, cronJob.Name)
 			if err != nil {
-				return nil, nil, updateStatus, err
+				return cronJob, nil, updateStatus, err
```
I changed the code following your suggestion here: https://github.com/kubernetes/kubernetes/pull/119776#discussion_r1311346986. And we no longer return `cronJobCopy` and `updateStatus` from `jm.syncCronJob()`, but rather accept the pointers to them as parameters, so there is now also no chance to mess them up by chance in this place by returning something different.

ASverdlov committed:
Return updateStatus bool instead of changing the reference

ASverdlov committed:
Explicitly ignore err in tests to fix linter

ASverdlov commented:
/retest

soltysh reviewed (approved):
/lgtm
/approve

ASverdlov commented:
/test pull-kubernetes-e2e-kind
